File: swift/Neetcode/LinkedList/Medium/19.RemoveNthNodeFromEnd.py
# Given the head of a linked list, remove the nth node from the end of the list 
# and return its head.

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def removeNthFromEnd(head, n):
    new_node = ListNode(val=0, next=head)
    left, right = new_node, head

    while n > 0 and right:
        right = right.next
        n -= 1

    while right:
        left = left.next
        right = right.next

    left.next = left.next.next
    return new_node.next

File: swift/Neetcode/LinkedList/Medium/19.RemoveNthNodeFromEnd.swift
// Given the head of a linked list, remove the nth node from the end of the list 
// and return its head.

func removeNthFromEnd(_ head: ListNode?, _ n: Int) -> ListNode? {
    let newHead = ListNode(0, head)
    var slow: ListNode? = newHead, fast = head

    for _ in 0..<n {
        fast = fast?.next
    }

    while fast != nil {
        slow = slow?.next
        fast = fast?.next
    }

    slow?.next = slow?.next?.next
    return newHead.next
}